### ISSUE TITLE
[question] Guidance related to my workflow

### ISSUE THREAD

**Issue by adnan-ali1** [conan-io/conan#15025]
### What is your question?

I have a few components which are in development stage continuously and I have a concept, I need some expert to have a look at the concept and guide me if this can be done in a more optimized way.

Dependencies tree
1. Base Component
2. Components using the Base component
3. Final Binary creation

e.g. 
basecomponent (upstream) -> component_a (downstream)
basecomponent (upstream) -> component_b (downstream)
basecomponent (upstream) -> component_c (downstream)

and then to build the final images
basecomponent (upstream) ->
component_a (upstream) ->
component_b (upstream) ->     FINAL_BINARY (downstream)
component_c (upstream) -> 

as I mentioned since all components are in continuous development stages here is what I am doing
1. Using version ranges in component_a, component_b and component_c for basecomponent to take the latest version from remote
2. Using version ranges in FINAL_BINARY, for basecomponent as well as the components a/b/c to take the latest version from remote

Build Details
1. During all the builds of components a/b/c, I generate a lockfile
2. Merge all the lockfile before creating running `conan create` on binary creation

Problem which I faced,
1. Since component a/b/c can be built at different time thus there pre-build packages can be possibily with different basecomponent
2. Now when I merge the conanfile and if component a/b/c are using different basecomponent versions they are downloaded as pre-built, reason is when I see the conan list for them and check the require the version is mentioned as e.g. '0.0.Z' and if basecomponent falls in this range then it does not build it again 
3. thus components with different versions of basecomponent are linked together

This I want to avoid.

How I (kindof) fixed it is , **AND HERE I NEED EXPERT OPINION** 
In components conanfile, I added 
```
    def package_id(self):
        self.info.requires["basecomponent"].full_version_mode()
```

Can this be better optimized ? or I am doing something wrong ?

### Have you read the CONTRIBUTING guide?

- [ ] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @adnan-ali1 

Thanks for your question.

> Since component a/b/c can be built at different time thus there pre-build packages can be possibily with different basecomponent

The key here is to use a lockfile as the input for those builds already. When you want to build them you don't want to build them in isolation, but instead as part of your final product build. Then, use the lockfile of the product dependencies, that will guarantee that all the components build against exactly the same base-component version and revision. 

The lockfile of the product can be evolved when you are already building ``base-component``. Using the product lockfile as input and creating a new lockfile from it, will add the new ``base-component`` version to the lockfile. Feeding this lockfile into the components will use the new version.

> def package_id(self):

What you are trying here is a different thing, it is about the binaries compatibility. But the problem you are seeing is before this, it is about resolving version.

In short, the lockfile-merge operations are more intended for merging different configurations or different products, but when you are trying to achieve a consistent dependency graph for one product, then the best approach is to use the "product" lockfile while building the intermediate packages of that dependency graph.

**Comment by memsharded** [comment]
Note: It is true that we are still missing some docs and examples for this kind of CI flows, it will be added in the new Devops guide when possible.

**Comment by adnan-ali1** [comment]
@memsharded I understand the approach which you described, however I am using the version ranges so that I can avoid it because

1. the package versions are growing dynamically with every new merge to release branch
2. computing ranges pre-hand is not so easy in this situation, because I want to build the binary with only the latest versions of all dependencies (and thus version ranges)

The whole CI pipeline looks like the following, for basecomponent and component
- Each component/basecomponent builds with there most recent dependency (_from all remotes_ < I think I have to change this behavior ) and are promoted to remote built
- From built remote packages pass unit-test criteria and are promoted to unittested remote
- Some more steps are done e.g. static code analysis etc, and each resulting in a promotion to new remote and eventually land in a remote e.g. pendingFunctionalTest
- A nightly job takes all packages (most latest) from pendingFunctionalTest remote and creates binary
- this binary is then tested under several hours long testing process.

Now you can see that `pendingFunctionalTest` might have even newer package e.g. of basecomponent as compared to the one which component a/b/c were built with in the first step or maybe in another case where the component was built with some basecomponent version which did not make it to the  `pendingFunctionalTest` remote. 

So, this is basically the concern and reason why I want to make sure that the packages are built with the same version of dependencies which made it to the specific remote. Do you think of a better way of obtaining this without the predefined lockfile per project because the product line is very big and mantaining such lockfiles would come with a lot of hassle and the dynamic version make it even hard to achieve. Would the workaround/fix which I wrote in my initial post make sense ?

**Comment by memsharded** [comment]
> So, this is basically the concern and reason why I want to make sure that the packages are built with the same version of dependencies which made it to the specific remote. Do you think of a better way of obtaining this without the predefined lockfile per project because the product line is very big and mantaining such lockfiles would come with a lot of hassle and the dynamic version make it even hard to achieve. Would the workaround/fix which I wrote in my initial post make sense ?

There are basically 3 approaches to the problem you are describing:

1. Making sure the different components that depends on the same ``base_component`` are built with the same verison of ``base_component``, to guarantee that when they are integrated, the precompiled binaries will match and integrate cleanly. 
2. Allow the different components to build with potentially different versions of ``base_component`` at their normal builds. Wait until the integration build fires, and then allow to link all of them together with a common version of ``base_component``, assuming it is binary compatible.
3. Same as 2, but if different versions are required than the ones that were used to build the component, then a re-build of the components will be forced too, to guarantee they are all built with the exact same version. 

3 and 1 achieve the same final result, in which all components are built with the same version of ``base_component``. Approach 1 did only 1 build, and approach 3 did 2 builds. Approach 3 also captures errors of building components against a specific base-component version much later, at night, when developers no longer have opportunity to fix it early.

Approach 2 is not that unfeasible and risky as it might sound. Actually, with the right versioning scheme, it could work pretty well. It doesn't matter against what specific patch version of a ``base_component`` dependency that is a static library is used, if the ``component`` is also a static library. As long as the public interface of ``base_component`` doesn't change, it should be safe, from the binary compatibility point of view.


If I understand your question correctly, your concerns are more in the line how to differentiate between 2 and 3, and how to control which packages needs to be re-built. The defaults of Conan 2.0 for ``package_id`` should be pretty safe, check:
- https://docs.conan.io/2/reference/binary_model.html
- https://docs.conan.io/2/reference/binary_model/dependencies.html. 
- https://docs.conan.io/2/reference/conanfile/attributes.html#reference-conanfile-attributes-package-id-modes

The Conan 2.0 defaults is even more strict than ``full_version_mode`` for the "embed" case. For the non-embed case, it is "minor_mode``, which means that it will require re-builds if dependencies change the major, but not if they change the patch. This might be a good convention to avoid unnecessary re-building when the public interface doesn't change.
If you want to change the global behavior, you don't need to change every recipe, you can configure that globally: https://docs.conan.io/2/reference/binary_model/custom_compatibility.html


With the new lockfiles in Conan 2.0, managing them should be much easier. But as everything in SW engineering, this is a tradeoff. If you go without lockfiles, you can delay the final integration and necessary re-buildings, but they will need to happen later, and that has its own pros and cons.

**Comment by adnan-ali1** [comment]
Thankyou @memsharded for the detailed response, 

I will look through the documents you shared. If there is something which I need to discuss further I will check with the slack channel or open another discussion. 

Closing this, thanks again @memsharded

**Comment by memsharded** [comment]
> I will look through the documents you shared. If there is something which I need to discuss further I will check with the slack channel or open another discussion.

The slack channel is great, and the community there is amazing, super helpful. But sometimes we don't have enough time to hang around there, if you want an "official" response, please don't hesitate to open a ticket here in Github, or if at some point you need to discuss something in private, you can write to info@conan.io. Thanks!
